classes/Address.php

```diff
@@ -245,14 +244,24 @@ public function delete()
      */
     protected function deleteCartAddress()
     {
+        $skipOrderSql = ' AND NOT EXISTS (SELECT 1 FROM ' . _DB_PREFIX_ . 'orders o WHERE o.`id_cart` = c.`id_cart`)';
+
         // keep pending carts, but unlink it from current address
-        $sql = 'UPDATE ' . _DB_PREFIX_ . 'cart
+        $sql = 'UPDATE ' . _DB_PREFIX_ . 'cart c
                     SET id_address_delivery = 0
-                    WHERE id_address_delivery = ' . $this->id;
+                    WHERE id_address_delivery = ' . $this->id . $skipOrderSql;
```

Comment:
In bug report https://github.com/PrestaShop/PrestaShop/issues/27340 you explained why `checkAndUpdateAddresses ()` needs to be improved but you did not explain this part?

Reply:
actually after thinking about it checkAndUpdateAddresses seems to be kind of a hack to fix the real issue: the id_address_delivery / id_address_invoice should have been properly reset in Address::delete at the first place.
With the original implementation, the id_address is only reset in case of a real delete (which occurs only if the address has no order attached to it). With my fix, it's called in all cases, but we now ignore all the cart attached to an order through this extra SQL test.
And I also properly reset all the place where id_address_delivery is used (i.e. in cart_product and customization).

At the very end, checkAndUpdateAddresses should be removed, but I decided to keep this sanitize check just in case.